packages/jest/src/migrations/update-17-1-0/move-options-to-target-defaults.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  createProjectGraphAsync,
+  formatFiles,
+  getProjects,
+  readNxJson,
+  TargetConfiguration,
+  Tree,
+  updateNxJson,
+  updateProjectConfiguration,
+} from '@nx/devkit';
+import { JestExecutorOptions } from '../../executors/jest/schema';
+import {
+  forEachExecutorOptions,
+  forEachExecutorOptionsInGraph,
+} from '@nx/devkit/src/generators/executor-options-utils';
+import { readTargetDefaultsForTarget } from 'nx/src/project-graph/utils/project-configuration-utils';
+
+export default async function update(tree: Tree) {
+  const nxJson = readNxJson(tree);
+
+  // Don't override anything if there are already target defaults for jest
+  if (nxJson.targetDefaults?.['@nx/jest:jest']) {
+    return;
+  }
+
+  nxJson.targetDefaults ??= {};
+
+  /**
+   * A set of targets which does not use any other executors
+   */
+  const jestTargets = new Set<string>();
+
+  const graph = await createProjectGraphAsync();
+
+  forEachExecutorOptionsInGraph(
+    graph,
+    '@nx/jest:jest',
+    (value, proj, targetName) => {
+      jestTargets.add(targetName);
+    }
+  );
+
+  // Workspace does not use jest?
+  if (jestTargets.size === 0) {
+    return;
+  }
+  // Use the project graph so targets which are inferred are considered
+  const projects = graph.nodes;
+  const projectMap = getProjects(tree);
+
+  const jestDefaults: TargetConfiguration<Partial<JestExecutorOptions>> =
+    (nxJson.targetDefaults['@nx/jest:jest'] = {});
+
+  // All jest targets have the same name
+  if (jestTargets.size === 1) {
+    const targetName = Array.from(jestTargets)[0];
+    if (nxJson.targetDefaults[targetName]) {
+      Object.assign(jestDefaults, nxJson.targetDefaults[targetName]);
+    }
+  }
+
+  jestDefaults.cache ??= true;
+
+  const inputs = ['default'];
+  inputs.push(nxJson.namedInputs?.production ? '^production' : '^default');
+  if (tree.exists('jest.preset.js')) {
+    inputs.push('{workspaceRoot}/jest.preset.js');
+  }
+  jestDefaults.inputs ??= inputs;
+
+  jestDefaults.options ??= {};
+  jestDefaults.options.passWithNoTests ??= true;
+  jestDefaults.configurations ??= {};
+  jestDefaults.configurations.ci ??= {};
+  jestDefaults.configurations.ci.ci ??= true;
+  jestDefaults.configurations.ci.codeCoverage ??= true;
+
+  // Cleanup old target defaults
+  for (const [targetDefaultKey, targetDefault] of Object.entries(
+    nxJson.targetDefaults
+  )) {
+    // If every target in every project does not use the target default anymore, it can be removed.
+    if (
+      Object.values(projects).every((p) =>
```

Comment:
It's a migration so not too worried about maintenance, but I find the long, nested `everyone` hard to grok. Would it better to use nested for-loops?

```js
for (const [targetDefaultKey, targetDefault] of Object.entries(
  nxJson.targetDefaults
)) {
  let isDefaultUsed = false;
  for (const p of Object.values(projects)) {
    for (const targetName of Object.keys(p.data?.targets ?? {})) {
      const defaults = readTargetDefaultsForTarget(
        targetName,
        nxJson.targetDefaults,
        projectMap.get(p.name).targets?.[targetName]?.executor
      );
      if (defaults === targetDefault) {
        isDefaultUsed = true;
        break;
      }
    }
  }
  if (!isDefaultUsed) {
    delete nxJson.targetDefaults[targetDefaultKey];
  }
}
```

Reply:
Yes, I agree.